server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -145,6 +145,9 @@ private RoutingNodes(RoutingTable routingTable, DiscoveryNodes discoveryNodes, b
                 }
             }
         }
+        for (var node : nodesToShards.values()) {
+            assert node.invariant();
+        }
```

Comment:
Invariant checking is expensive in case of big collections so it is verified once in the end instead of after every operation.

Reply:
That makes sense. One suggestion: I think we do the invariant check in one line with method references:

```
nodesToShards.values().forEach(RoutingNode::invariant)
```

Reply:
Link https://github.com/elastic/elasticsearch/pull/88951 :)